mypy/newsemanal/semanal.py

```diff
@@ -2803,6 +2803,10 @@ def process_typevar_parameters(self, args: List[Expression],
     def basic_new_typeinfo(self, name: str, basetype_or_fallback: Instance) -> TypeInfo:
         class_def = ClassDef(name, Block([]))
         class_def.fullname = self.qualified_name(name)
+        if self.is_func_scope() and not self.type:
+            # Full names of generated classes should always be prefixed with the module names
+            # even if they are nested in a function, since these classes will be (de-)serialized.
+            class_def.fullname = self.cur_mod_id + '.' + class_def.fullname
```

Comment:
It looks like this can collide with another class name. Maybe you should use the `@line` suffix that we use elsewhere?

Reply:
This should be already done by class generation code. I will add a comment.